runtime/darwinia/src/messages/crab_messages.rs

```diff
@@ -0,0 +1,276 @@
+};
+use bp_runtime::{messages::DispatchFeePayment, ChainId};
+use bridge_runtime_common::messages::{
+	self,
+	source::{self, FromBridgedChainMessagesDeliveryProof, FromThisChainMessagePayload},
+	target::{
+		self, FromBridgedChainEncodedMessageCall, FromBridgedChainMessageDispatch,
+		FromBridgedChainMessagePayload, FromBridgedChainMessagesProof,
+	},
+	MessageBridge, MessageTransaction,
+};
+use frame_support::{
+	weights::{DispatchClass, Weight},
+	RuntimeDebug,
+};
+use pallet_bridge_messages::EXPECTED_DEFAULT_MESSAGE_LENGTH;
+use sp_runtime::{traits::Zero, FixedPointNumber, FixedU128, MultiSignature, MultiSigner};
+use sp_std::{convert::TryFrom, ops::RangeInclusive};
+// --- darwinia-network ---
+use crate::*;
+use dp_s2s::{CallParams, CreatePayload};
+
+/// Message payload for Darwinia -> Crab messages.
+pub type ToCrabMessagePayload = FromThisChainMessagePayload<WithCrabMessageBridge>;
+/// The s2s issuing pallet index in the crab chain runtime
+pub const CRAB_S2S_ISSUING_PALLET_INDEX: u8 = 49;
+
+#[derive(RuntimeDebug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct ToCrabOutboundPayload;
+
+impl CreatePayload<AccountId, MultiSigner, MultiSignature> for ToCrabOutboundPayload {
+	type Payload = ToCrabMessagePayload;
+
+	fn create(
+		origin: CallOrigin<AccountId, MultiSigner, MultiSignature>,
+		spec_version: u32,
+		weight: u64,
+		call_params: CallParams,
+		dispatch_fee_payment: DispatchFeePayment,
+	) -> Result<Self::Payload, &'static str> {
+		let call = Self::encode_call(CRAB_S2S_ISSUING_PALLET_INDEX, call_params)?;
+		return Ok(ToCrabMessagePayload {
+			spec_version,
+			weight,
+			origin,
+			call,
+			dispatch_fee_payment,
+		});
+	}
+}
+
+/// Message verifier for Darwinia -> Crab messages.
+pub type ToCrabMessageVerifier<R> = DarwiniaFromThisChainMessageVerifier<WithCrabMessageBridge, R>;
+/// Message payload for Crab -> Darwinia messages.
+pub type FromCrabMessagePayload = FromBridgedChainMessagePayload<WithCrabMessageBridge>;
+/// Encoded Darwinia Call as it comes from Crab.
+pub type FromCrabEncodedCall = FromBridgedChainEncodedMessageCall<Call>;
+/// Messages proof for Crab -> Darwinia messages.
+type FromCrabMessagesProof = FromBridgedChainMessagesProof<Hash>;
+/// Messages delivery proof for Darwinia -> Crab messages.
+type ToCrabMessagesDeliveryProof = FromBridgedChainMessagesDeliveryProof<Hash>;
+/// Call-dispatch based message dispatch for Crab -> Darwinia messages.
+pub type FromCrabMessageDispatch =
+	FromBridgedChainMessageDispatch<WithCrabMessageBridge, Runtime, Ring, S2sBridgeDispatch>;
+
+/// Initial value of `CrabToDarwiniaConversionRate` parameter.
+pub const INITIAL_CRAB_TO_DARWINIA_CONVERSION_RATE: FixedU128 =
+	FixedU128::from_inner(FixedU128::DIV);
+
+frame_support::parameter_types! {
+	/// Crab to Darwinia conversion rate. Initially we treat both tokens as equal.
+	pub storage CrabToDarwiniaConversionRate: FixedU128 = INITIAL_CRAB_TO_DARWINIA_CONVERSION_RATE;
+}
```

Comment:
Used in estimate fee, @fewensa need you to check we use this in bridger.

Reply:
Plan to remove https://github.com/darwinia-network/bridger/pull/321